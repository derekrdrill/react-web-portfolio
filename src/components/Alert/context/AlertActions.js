export const handleAlert = (msg, title, type, alertDispatch) => {
  alertDispatch({
    type: 'SET_ALERT',
    fadeOut: false,
    payload: { msg: msg, title: title, type: type },
  });

  setTimeout(() => alertDispatch({ type: 'FADE_ALERT', fadeOut: true }), 2000);
  setTimeout(() => alertDispatch({ type: 'REMOVE_ALERT' }), 2400);
};
