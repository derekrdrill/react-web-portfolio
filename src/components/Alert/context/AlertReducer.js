export const alertReducer = (state, action) => {
  switch (action.type) {
    case 'SET_ALERT':
      return {
        ...state,
        alert: action.payload,
        fadeOut: action.fadeOut,
      };
    case 'FADE_ALERT':
      return {
        ...state,
        fadeOut: action.fadeOut,
      };
    case 'REMOVE_ALERT':
      return {
        ...state,
        alert: null,
        fadeOut: false,
      };
    default:
      return state;
  }
};
