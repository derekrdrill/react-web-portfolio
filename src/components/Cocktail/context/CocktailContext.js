import React, { createContext, useReducer } from 'react';
import PropTypes from 'prop-types';

import { cocktailReducer } from './CocktailReducer';

export const CocktailContext = createContext();

const CocktailProvider = ({ children }) => {
  const initialState = {
    advancedSearch: false,
    loading: false,
    cocktails: [],
    ingredients: [],
    searchType: 'name',
    searchResults: null,
    selectedIngredients: [],
  };

  const [state, cocktailDispatch] = useReducer(cocktailReducer, initialState);

  return (
    <CocktailContext.Provider
      value={{
        cocktailDispatch,
        ...state,
      }}
    >
      {children}
    </CocktailContext.Provider>
  );
};

CocktailProvider.propTypes = {
  children: PropTypes.node,
};

export default CocktailProvider;
