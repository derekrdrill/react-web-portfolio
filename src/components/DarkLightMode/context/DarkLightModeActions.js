export const handleSetDarkMode = (darkMode, darkLightModeDispatch) => {
  localStorage.setItem('darkMode', JSON.stringify(!darkMode));

  darkLightModeDispatch({
    type: 'SET_DARK_MODE',
    darkMode: !darkMode,
  });
};
