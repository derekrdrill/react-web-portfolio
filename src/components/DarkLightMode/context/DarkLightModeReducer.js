export const darkLightModeReducer = (state, action) => {
  switch (action.type) {
    case 'SET_DARK_MODE':
      return {
        ...state,
        darkMode: action.darkMode,
      };
    default:
      return state;
  }
};
