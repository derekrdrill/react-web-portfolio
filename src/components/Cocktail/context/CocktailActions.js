import axios from 'axios';

import { handleAlert } from '../../Alert/context/AlertActions';

const COCKTAIL_SEARCH_URL = process.env.REACT_APP_COCKTAIL_SEARCH_URL;
const COCKTAIL_LIST_URL = process.env.REACT_APP_COCKTAIL_LIST_URL;
const COCKTAIL_KEY = process.env.REACT_APP_COCKTAIL_KEY;
const COCKTAIL_HOST = process.env.REACT_APP_COCKTAIL_HOST;

export const getAllCocktails = async cocktailDispatch => {
  const options = {
    method: 'GET',
    url: COCKTAIL_SEARCH_URL,
    params: { s: '' },
    headers: {
      'X-RapidAPI-Key': COCKTAIL_KEY,
      'X-RapidAPI-Host': COCKTAIL_HOST,
    },
  };

  axios
    .request(options)
    .then(function (response) {
      cocktailDispatch({
        type: 'SET_COCKTAILS',
        cocktails: response.data.drinks.filter(drink => !drink.strDrink.includes('1-900-FUK')),
      });
    })
    .catch(function (error) {
      console.error(error);
    });
};

export const getAllGlasses = async cocktailDispatch => {
  const options = {
    method: 'GET',
    url: COCKTAIL_LIST_URL,
    params: { g: 'list' },
    headers: {
      'X-RapidAPI-Key': COCKTAIL_KEY,
      'X-RapidAPI-Host': COCKTAIL_HOST,
    },
  };

  axios
    .request(options)
    .then(function (response) {
      cocktailDispatch({
        type: 'SET_GLASSES',
        glasses: response.data.drinks,
      });
    })
    .catch(function (error) {
      console.error(error);
    });
};

export const getAllIngredients = cocktailDispatch => {
  const options = {
    method: 'GET',
    url: COCKTAIL_LIST_URL,
    params: { i: 'list' },
    headers: {
      'X-RapidAPI-Key': COCKTAIL_KEY,
      'X-RapidAPI-Host': COCKTAIL_HOST,
    },
  };

  axios
    .request(options)
    .then(function (response) {
      cocktailDispatch({
        type: 'SET_INGREDIENTS',
        ingredients: response.data.drinks,
      });
    })
    .catch(function (error) {
      console.error(error);
    });
};

export const getSearchOptions = (searchType, cocktails, ingredients, glasses) =>
  searchType === 'name' && cocktails.length > 0
    ? cocktails.map(option => option.strDrink).sort()
    : searchType === 'ingredients' && ingredients.length > 0
    ? ingredients.map(option => option.strIngredient1).sort()
    : searchType === 'glass' && glasses.length > 0
    ? glasses.map(option => option.strGlass).sort()
    : [];

export const getCocktailByName = async (name, cocktailDispatch) => {
  const options = {
    method: 'GET',
    url: COCKTAIL_SEARCH_URL,
    params: { s: name },
    headers: {
      'X-RapidAPI-Key': COCKTAIL_KEY,
      'X-RapidAPI-Host': COCKTAIL_HOST,
    },
  };

  if (!name) {
    cocktailDispatch({
      type: 'SET_SEARCH_RESULTS',
      searchResults: null,
    });

    cocktailDispatch({
      type: 'SET_SEARCH_RESULTS_LENGTH',
      searchResultsLength: 0,
    });
  } else {
    axios
      .request(options)
      .then(function (response) {
        const drinks = response.data.drinks;

        cocktailDispatch({
          type: 'SET_SEARCH_RESULTS',
          searchResults: drinks,
        });

        cocktailDispatch({
          type: 'SET_SEARCH_RESULTS_LENGTH',
          searchResultsLength: drinks.length,
        });
      })
      .catch(function (error) {
        console.error(error);
      });
  }
};

export const getCocktailsByIngredientOrGlass = async (
  alertDispatch,
  cocktailDispatch,
  cocktails,
  searchType,
  selectedIngredients,
  searchData,
) => {
  let newCocktails = [];
  let newCocktailsNoDups = [];

  await cocktails.forEach(cocktail => {
    if (searchType === 'ingredients') {
      let foundIngredients = [];

      selectedIngredients.forEach(selectedIngredient => {
        for (let i = 1; i <= 15; i++) {
          let ingredient =
            cocktail[`strIngredient${i}`] && cocktail[`strIngredient${i}`].toUpperCase();

          if (ingredient && ingredient.includes(selectedIngredient.toUpperCase())) {
            foundIngredients = [...foundIngredients, ...[ingredient]];
          }
        }
      });

      if (foundIngredients.length >= selectedIngredients.length) {
        let fullMatch = false;

        foundIngredients.forEach(foundIngredient => {
          selectedIngredients.forEach(selectedIngredient => {
            if (foundIngredient.includes(selectedIngredient.toUpperCase())) {
              fullMatch = true;
            } else {
              fullMatch = false;
            }
          });
        });

        if (fullMatch) {
          newCocktails = [...newCocktails, ...[cocktail]];
        }
      }
    } else {
      if (searchData === cocktail.strGlass) {
        newCocktails = [...newCocktails, ...[cocktail]];
      }
    }
  });

  newCocktailsNoDups = newCocktails.filter(
    (v, i, a) => a.findIndex(v2 => JSON.stringify(v) === JSON.stringify(v2)) === i,
  );

  if (newCocktailsNoDups.length > 0) {
    cocktailDispatch({
      type: 'SET_SEARCH_RESULTS',
      searchResults: newCocktailsNoDups.sort((a, b) =>
        a.strDrink < b.strDrink ? -1 : a.strDrink > b.strDrink ? 1 : 0,
      ),
    });

    cocktailDispatch({
      type: 'SET_SEARCH_RESULTS_LENGTH',
      searchResultsLength: newCocktailsNoDups.length,
    });
  } else {
    if (searchData) {
      handleAlert('No drinks found with those indredients', 'Try again', 'warning', alertDispatch);
    }

    cocktailDispatch({
      type: 'SET_SEARCH_RESULTS',
      searchResults: [],
    });

    cocktailDispatch({
      type: 'SET_SEARCH_RESULTS_LENGTH',
      searchResultsLength: 0,
    });
  }
};

export const handleSearchBarChange = async (
  alertDispatch,
  cocktailDispatch,
  searchData,
  searchType,
  cocktails,
) => {
  if (searchType === 'name') {
    getCocktailByName(searchData, cocktailDispatch);
  } else if (searchType === 'glass') {
    getCocktailsByIngredientOrGlass(
      alertDispatch,
      cocktailDispatch,
      cocktails,
      searchType,
      null,
      searchData,
    );
  } else {
    let newSearchData =
      searchData.length > 3
        ? searchData.filter((searchItem, searchItemKey) => searchItemKey !== 2)
        : searchData;

    cocktailDispatch({
      type: 'SET_SELECTED_INGREDIENTS',
      selectedIngredients: newSearchData,
    });
  }
};

export const setSearchType = (e, cocktailDispatch) => {
  cocktailDispatch({
    type: 'SET_SEARCH_TYPE',
    searchType: e.target.value,
  });

  cocktailDispatch({
    type: 'SET_SEARCH_RESULTS',
    searchResults: null,
  });

  cocktailDispatch({
    type: 'SET_SEARCH_RESULTS_LENGTH',
    searchResultsLength: 0,
  });

  cocktailDispatch({
    type: 'SET_SELECTED_INGREDIENTS',
    selectedIngredients: [],
  });
};
