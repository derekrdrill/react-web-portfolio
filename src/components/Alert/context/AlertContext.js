import React, { createContext, useReducer } from 'react';
import { alertReducer } from './AlertReducer';

export const AlertContext = createContext();

export const AlertProvider = ({ children }) => {
  const initialState = {
    alert: null,
    fadeOut: false,
  };

  const [state, alertDispatch] = useReducer(alertReducer, initialState);

  return <AlertContext.Provider value={{ ...state, alertDispatch }}>{children}</AlertContext.Provider>;
};
