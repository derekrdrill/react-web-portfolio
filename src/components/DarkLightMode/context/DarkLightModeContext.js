import React, { createContext, useReducer } from 'react';
import PropTypes from 'prop-types';
import { darkLightModeReducer } from './DarkLightModeReducer';

export const DarkLightModeContext = createContext();

export const DarkLightModeProvider = ({ children }) => {
  const initialState = {
    darkMode: true,
  };

  const [state, darkLightModeDispatch] = useReducer(darkLightModeReducer, initialState);

  return (
    <DarkLightModeContext.Provider value={{ ...state, darkLightModeDispatch }}>
      {children}
    </DarkLightModeContext.Provider>
  );
};

DarkLightModeProvider.propTypes = {
  children: PropTypes.node,
};
