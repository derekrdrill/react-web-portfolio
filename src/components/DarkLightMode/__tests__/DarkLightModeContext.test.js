import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DarkLightModeProvider } from '../context/DarkLightModeContext';

const renderer = new ShallowRenderer();

describe('DarkLightModeProvider with me tests', () => {
  it('renders correctly', () => {
    renderer.render(<DarkLightModeProvider />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
