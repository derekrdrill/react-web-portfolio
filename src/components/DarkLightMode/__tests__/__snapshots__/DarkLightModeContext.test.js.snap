// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DarkLightModeProvider with me tests renders correctly 1`] = `
<Context.Provider
  value={
    Object {
      "darkLightModeDispatch": [Function],
      "darkMode": true,
    }
  }
/>
`;
